### PR TITLE
RELATED: SD-970 typing for post event filter context

### PR DESCRIPTION
**Pull request by lanhq**

### CODE REVIEW

**Comment by yenkins** [comment]
Currently destined for [release R208](https://confluence.intgdc.com/display/Dev/R208).
This release is **zero downtime** release.
Code drop: 2020-07-02 00:00.

**Comment by yenkins** [comment]
Story [SD-970](http://jira.intgdc.com/browse/SD-970) Control filters - Low level message handler
	Status: Hacking
	Priority: Major
	Assignee: Lan Huynh

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-pull-request-compliance-check/109/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-unit-tests-zuul-docker-el7/191/ : FAILURE in 37s

File: src/embedding/kd.ts [hunk 1]
```diff
@@ -332,6 +358,53 @@ export namespace EmbeddedKpiDashboard {
...
+     * Data type of SetFilterContext command
+     */
+    export type SetFilterContextCommandData = IGdcKdMessageEnvelope<GdcKdCommandType.SetFilterContext, EmbeddedGdc.IFilterContextContent>;
+
+    /**
+     * Add or update the filter context
+     *
+     * Contract:
+     * - if filters is same with filters on the KD filter bar, then update the filters on the filter bar and apply the filters to dashboard
+     * - if filters is new one, then add it to the KD filter bar and apply to dashboard
```
Comment (ptquang86) [inline src/embedding/kd.ts:371]: 1. `filters are...`
2. `... add them ...`. Note that this is for KD edit mode only

File: src/embedding/kd.ts [hunk 2]
```diff
@@ -332,6 +358,53 @@ export namespace EmbeddedKpiDashboard {
...
+     * - in-case the AD can not apply the filters, a CommandFailed will be posted. The reason could be:
+     *   - exceed the limit number of filter items for selected filter values
```
Comment (ptquang86) [inline src/embedding/kd.ts:373]: 1. `KD`
2. `exceed the limit number of filter items for dashboard`

File: src/embedding/kd.ts [hunk 3]
```diff
@@ -332,6 +358,53 @@ export namespace EmbeddedKpiDashboard {
...
+     */
+    export type SetFilterContextCommand = IGdcKdMessageEvent<GdcKdCommandType.SetFilterContext, EmbeddedGdc.IFilterContextContent>;
+
+    /**
+     * Type-guard checking whether an object is an instance of {@link SetFilterContextCommand}
+     *
+     * @param obj - object to test
+     */
+    export function isSetFilterContextCommandData(obj: any): obj is SetFilterContextCommandData {
```
Comment (ptquang86) [inline src/embedding/kd.ts:382]: is it `isSetFilterContextCommand`?

File: src/embedding/kd.ts [hunk 4]
```diff
@@ -332,6 +358,53 @@ export namespace EmbeddedKpiDashboard {
...
+    /**
+     * Data type of removeFilterContext command
+     */
+    export type RemoveFilterContextCommandData = IGdcKdMessageEnvelope<GdcKdCommandType.RemoveFilterContext, EmbeddedGdc.IFilterContextContent>;
+
+    /**
+     * Remove the filter context
+     * Contract:
+     * - if filters is in the filter bar, then update the filters on the filter bar and apply the filters to insight
+     * - if filters is not in the filter bar, then a CommandFailed will be posted.
```
Comment (ptquang86) [inline src/embedding/kd.ts:395]: 1. `filters are ...`
2. `then remove the filters on the filter bar and update insight`

File: src/embedding/kd.ts [hunk 5]
```diff
@@ -332,6 +358,53 @@ export namespace EmbeddedKpiDashboard {
...
+     */
+    export type RemoveFilterContextCommand = IGdcKdMessageEvent<GdcKdCommandType.RemoveFilterContext, EmbeddedGdc.IFilterContextContent>;
+
+    /**
+     * Type-guard checking whether an object is an instance of {@link RemoveFilterContextCommand}
+     *
+     * @param obj - object to test
+     */
+    export function isRemoveFilterContextCommandData(obj: any): obj is RemoveFilterContextCommandData {
```
Comment (ptquang86) [inline src/embedding/kd.ts:404]: is it `isRemoveFilterContextCommand`?

File: src/embedding/kd.ts [hunk 6]
```diff
@@ -618,4 +691,51 @@ export namespace EmbeddedKpiDashboard {
      * This event is emitted after dashboard has been exported to PDF
      */
     export type ExportToPdfFinishedData = IGdcKdMessageEnvelope<GdcKdEventType.ExportedToPdf, ExportToPdfFinishedBody>;
+
+    //
+    // setFilterContext finished
+    //
+
+    /**
+     * Data type of event that was emitted after finish set filter context
```
Comment (ptquang86) [inline src/embedding/kd.ts:700]: `... after finishing ...`

File: src/embedding/kd.ts [hunk 7]
```diff
@@ -618,4 +691,51 @@ export namespace EmbeddedKpiDashboard {
...
+        GdcKdEventType.SetFilterContextFinished,
+        IKdAvailableCommands
+    >;
+
+    //
+    // removeFilterContext finished
+    //
+
+    /**
+     * Data type of event that was emitted after finish set filter context
```
Comment (ptquang86) [inline src/embedding/kd.ts:714]: `... after finishing remove ...`

File: src/embedding/kd.ts [hunk 8]
```diff
@@ -618,4 +691,51 @@ export namespace EmbeddedKpiDashboard {
...
+    // FilterContext changed
+    //
+
+    /**
+     * Main data of Filter context changed event
+     */
+    export type FilterContextChangedBody = IKdAvailableCommands & EmbeddedGdc.IFilterContextContent;
+
+    /**
+     * Data type of event that was emitted after finish set filter context
```
Comment (ptquang86) [inline src/embedding/kd.ts:733]: `... after finishing change ...`

File: src/embedding/ad.ts [hunk 9]
```diff
@@ -516,6 +542,54 @@ export namespace EmbeddedAnalyticalDesigner {
...
+
+    /**
+     * Add or update the filter context
+     *
+     * Contract:
+     * - if filters is same with filters on the AD filter bar, then update the filters on the filter bar and apply the filters to insight
+     * - if filters is new one, then add it to the AD filter bar and apply to insight
+     * - in-case the AD can not apply the filters, a CommandFailed will be posted. The reason could be
+     *   - add new filter in view mode
+     *   - exceed the limit number of filter items for selected filter values
```
[the remark this answers is not in the document]
Comment (lvkhanh) [reply src/embedding/ad.ts:558]: For AD `CommandFailed` will be posted in cases:
- filter is not existed
- exceed the limit number of filter items
- filter is existed but wrong elements input data

File: src/embedding/ad.ts [hunk 10]
```diff
@@ -516,6 +542,54 @@ export namespace EmbeddedAnalyticalDesigner {
...
+     * Data type of SetFilterContext command
+     */
+    export type SetFilterContextCommandData = IGdcAdMessageEnvelope<GdcAdCommandType.SetFilterContext, EmbeddedGdc.IFilterContextContent>;
+
+    /**
+     * Add or update the filter context
+     *
+     * Contract:
+     * - if filters is same with filters on the AD filter bar, then update the filters on the filter bar and apply the filters to insight
+     * - if filters is new one, then add it to the AD filter bar and apply to insight
```
[the remark this answers is not in the document]
Comment (lvkhanh) [reply src/embedding/ad.ts:555]:
```
if filters existed on the AD filter bar, then the filters will be updated and the insight will be applied
if filters are new one, then the the filters will be added and the insight will be applied
```

File: src/embedding/ad.ts [hunk 11]
```diff
@@ -516,6 +542,54 @@ export namespace EmbeddedAnalyticalDesigner {
...
+    /**
+     * Data type of removeFilterContext command
+     */
+    export type RemoveFilterContextCommandData = IGdcAdMessageEnvelope<GdcAdCommandType.RemoveFilterContext, EmbeddedGdc.IFilterContextContent>;
+
+    /**
+     * Remove the filter context
+     * Contract:
+     * - if filters is in the filter bar, then update the filters on the filter bar and apply the filters to insight
+     * - if filters is not in the filter bar, then a CommandFailed will be posted.
```
[the remark this answers is not in the document]
Comment (lvkhanh) [reply src/embedding/ad.ts:580]:
```
if filters existed on the AD filter bar, then the filters will be removed and the insight will be updated
if filters are in the AD filter bar, then a CommandFailed will be posted.
```

File: src/embedding/kd.ts (hunk 1 above)
Comment (lvkhanh) [reply src/embedding/kd.ts:371]:
```
if filters existed on the KD filter bar, then the filters will be updated and the insight on dashboard will be applied
if filters are new one, then the the filters will be added and the insight on dashboard will be applied
```

File: src/embedding/kd.ts (hunk 2 above)
Comment (lvkhanh) [reply src/embedding/kd.ts:373]:
```
For KD `CommandFailed` will be posted in cases:
filter is not existed
exceed the limit number of filter items
filter is existed but wrong elements input data
```

File: src/embedding/kd.ts (hunk 4 above)
Comment (lvkhanh) [reply src/embedding/kd.ts:395]:
```
if filters existed on the KD filter bar, then the filters will be removed and the insight will be updated
if filters are in the KD filter bar, then a CommandFailed will be posted.
```

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-pull-request-compliance-check/110/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-unit-tests-zuul-docker-el7/192/ : FAILURE in 45s

**Comment by yenkins** [comment]
Currently destined for [release R](https://confluence.intgdc.com/display/Dev/R).
This release is **zero downtime** release.
Code drop: .

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-pull-request-compliance-check/111/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-unit-tests-zuul-docker-el7/193/ : SUCCESS in 35s

**Comment by lvkhanh** [comment]
LGTM

File: src/embedding/kd.ts [hunk 12]
```diff
@@ -332,6 +358,69 @@ export namespace EmbeddedKpiDashboard {
...
+     */
+    export type RemoveFilterContextCommandData = IGdcKdMessageEnvelope<
+        GdcKdCommandType.RemoveFilterContext,
+        EmbeddedGdc.IFilterContextContent
+    >;
+
+    /**
+     * Remove the filter context
+     * Contract:
+     * - if filters are in the filter bar, then update the filters on the filter bar and apply the filters to insight
```
Comment (ptquang86) [inline src/embedding/kd.ts:407]: `then remove the filters on the filter bar and update insight`

File: src/embedding/ad.ts [hunk 13]
```diff
@@ -516,6 +542,68 @@ export namespace EmbeddedAnalyticalDesigner {
...
+        EmbeddedGdc.IFilterContextContent
+    >;
+
+    /**
+     * Add or update the filter context
+     *
+     * Contract:
+     * - if filters are same with filters on the AD filter bar, then update the filters on the filter bar
+     *   and apply the filters to insight
+     * - if filters are new one, then add it to the AD filter bar and apply to insight
```
Comment (ptquang86) [inline src/embedding/ad.ts:559]: `are new, then add them to...`

File: src/embedding/common.ts [hunk 14]
```diff
@@ -217,3 +219,23 @@ export interface IDrillableItemsCommandBody extends ISimpleDrillableItemsCommand
...
+    export type IPositiveAttributeFilter = AFM.IPositiveAttributeFilter;
+    export type INegativeAttributeFilter = AFM.INegativeAttributeFilter;
+    export type IAbsoluteDateFilter = AFM.IAbsoluteDateFilter;
+    export type IRelativeDateFilter = AFM.IRelativeDateFilter;
+    export type AttributeFilterItem = IPositiveAttributeFilter | INegativeAttributeFilter;
+    export type DateFilterItem = IAbsoluteDateFilter | IRelativeDateFilter;
+    export type FilterItem = DateFilterItem | AttributeFilterItem;
+
+    /**
+     * The filter context content that used to exchange the filter context
```
Comment (ptquang86) [inline src/embedding/common.ts:234]: `... that is used to ...`

File: src/embedding/kd.ts [hunk 15]
```diff
@@ -332,6 +358,69 @@ export namespace EmbeddedKpiDashboard {
...
+        EmbeddedGdc.IFilterContextContent
+    >;
+
+    /**
+     * Add or update the filter context
+     *
+     * Contract:
+     * - If filters are same with filters on the KD filter bar, then update the filters on the filter bar
+     *   and apply the filters to dashboard
+     * - In edit mode, if filters are new one and then add it to the KD filter bar and apply to dashboard
```
Comment (ptquang86) [inline src/embedding/kd.ts:375]: `... are new and then add them to ...`

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-pull-request-compliance-check/112/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-unit-tests-zuul-docker-el7/194/ : SUCCESS in 56s

**Comment by ptquang86** [comment]
LGTM

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-pull-request-compliance-check/113/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-unit-tests-zuul-docker-el7/195/ : SUCCESS in 55s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-pull-request-compliance-gate/40/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-typings-unit-tests-zuul-docker-el7/196/ : SUCCESS in 1m 01s
- https://checklist.intgdc.com/job/compliance/job/gooddata-typings-fossa-scanning-zuul-el7/4/ : SUCCESS in 14s (non-voting)

**Pull request merged**